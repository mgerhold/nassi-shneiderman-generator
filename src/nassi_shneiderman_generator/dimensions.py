from typing import final


@final
class Dimensions:
    _width: float
    _height: float
    _depth: float

    def __init__(self, width: float, height: float, depth: float) -> None:
        self._width = width
        self._height = height
        self._depth = depth

    @property
    def width(self) -> float:
        return self._width

    @property
    def height(self) -> float:
        return self._height

    @property
    def depth(self) -> float:
        return self._depth

    @property
    def total_height(self) -> float:
        return self.height + self.depth
