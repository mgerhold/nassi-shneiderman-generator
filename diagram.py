from typing import final

from symbols import Symbol


@final
class Diagram:
    def __init__(self, contents: Symbol) -> None:
        self._contents = contents

    def emit(self) -> str:
        result = "\\begin{tikzpicture}\n"
        result += "\\end{tikzpicture}\n"
        return result
