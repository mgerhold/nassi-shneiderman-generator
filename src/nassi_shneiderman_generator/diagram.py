from typing import final, Optional, Final

from nassi_shneiderman_generator.symbols import Symbol


@final
class Diagram:
    def __init__(self, contents: Symbol, min_width: float = 80.0) -> None:
        self._contents = contents
        self._min_width = min_width

    def emit(self) -> str:
        result = "\\begin{tikzpicture}\n"
        required_size: Final = self._contents.required_size
        result += self._contents.emit(
            position=(0, 0),
            size=(max(self._min_width, required_size[0]), required_size[1]),
        )
        result += "\\end{tikzpicture}\n"
        return result
