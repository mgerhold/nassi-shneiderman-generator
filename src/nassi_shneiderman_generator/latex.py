from functools import cache
from typing import Final
import subprocess
import tempfile
import re
from pathlib import Path
import cv2

from nassi_shneiderman_generator.dimensions import Dimensions


@cache
def measure_latex_dimensions(latex_code: str) -> Dimensions:
    file_contents: Final = rf"""\documentclass{{article}}
    \usepackage[margin=0pt,paperwidth=50cm,paperheight=50cm]{{geometry}}
    \usepackage{{calc}}
    \pagestyle{{empty}}

    \begin{{document}}
        \newsavebox{{\mybox}}
        \savebox{{\mybox}}{{{latex_code}}}
        \typeout{{DIMENSIONS: \the\wd\mybox\space\the\ht\mybox\space\the\dp\mybox}}
        \usebox{{\mybox}}
    \end{{document}}
"""

    with tempfile.TemporaryDirectory() as temp_dir:
        tex_file: Final = Path(temp_dir) / "measure.tex"
        tex_file.write_text(file_contents, encoding="utf-8")

        try:
            result = subprocess.run(
                ["pdflatex", "-interaction=nonstopmode", str(tex_file)],
                cwd=temp_dir,
                capture_output=True,
                text=True,
                timeout=30,
            )

            pattern: Final = r"DIMENSIONS: ([0-9.]+)pt ([0-9.]+)pt ([0-9.]+)pt"
            match = re.search(pattern, result.stdout)

            if match:
                width = float(match.group(1))
                height = float(match.group(2))
                depth = float(match.group(3))
                return Dimensions(
                    width=width,
                    height=height,
                    depth=depth,
                )
            raise RuntimeError("Failed to extract dimensions from LaTeX output.")
        except subprocess.TimeoutExpired:
            raise RuntimeError("pdflatex timed out")
        except FileNotFoundError:
            raise RuntimeError("pdflatex is not installed or not found in PATH")


def _show_image(path: str) -> None:
    img = cv2.imread(path, cv2.IMREAD_UNCHANGED)
    cv2.imshow("LaTeX Output", img)  # type: ignore
    cv2.waitKey(0)
    cv2.destroyAllWindows()


def render_latex_to_pdf(latex_body: str, target_filepath: Path) -> None:
    full_tex = (
        r"""\documentclass[tikz]{standalone}
\usepackage{tikz}
\begin{document}
"""
        + latex_body
        + r"""
    \end{document}
    """
    )

    with tempfile.TemporaryDirectory() as tmpdir:
        tmp_path = Path(tmpdir).resolve()
        tex_file = tmp_path / "doc.tex"
        tex_file.parent.mkdir(parents=True, exist_ok=True)
        tex_file.write_text(full_tex, encoding="utf-8")

        subprocess.run(
            ["pdflatex", "-interaction=nonstopmode", str(tex_file)],
            cwd=tmp_path,
            check=True,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
        pdf_file = tmp_path / "doc.pdf"
        pdf_file.rename(target_filepath)


def render_latex_and_show(latex_body: str):
    # Create temp directory
    with tempfile.TemporaryDirectory() as tmpdir:
        tmp_path = Path(tmpdir).resolve()
        pdf_file = tmp_path / "doc.pdf"
        png_file = tmp_path / "doc.png"  # output of pdftoppm

        render_latex_to_pdf(latex_body, pdf_file)

        # Convert PDF to PNG
        try:
            subprocess.run(
                [
                    "pdftoppm",
                    "-png",
                    "-singlefile",
                    str(pdf_file),
                    str(tmp_path / "doc"),
                ],
                check=True,
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
            )
        except subprocess.CalledProcessError:
            print("PDF to PNG conversion failed.")
            return

        # Open image
        try:
            _show_image(str(png_file))
        except Exception as e:
            print("Failed to open image:", e)
