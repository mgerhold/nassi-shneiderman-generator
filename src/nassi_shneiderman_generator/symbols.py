from enum import Enum, auto
from typing import NamedTuple
from typing import override
from typing import final, Final
from abc import ABC, abstractmethod
from typing import Optional

from nassi_shneiderman_generator.dimensions import Dimensions
from nassi_shneiderman_generator.latex import measure_latex_dimensions

_TEXT_MARGIN = 4.0


def _lerp(start: float, end: float, t: float) -> float:
    return start + (end - start) * t


def _options(color: Optional[str]) -> str:
    return "" if color is None else f"[{color}]"


def _line(x1: float, y1: float, x2: float, y2: float) -> str:
    return rf"    \draw ({x1}pt, {y1}pt) -- ({x2}pt, {y2}pt);" + "\n"


def _rectangle(x1: float, y1: float, x2: float, y2: float, color: Optional[str]) -> str:
    return (
        rf"    \draw{_options(color)} ({x1}pt,{y1}pt) rectangle ({x2}pt, {y2}pt);"
        + "\n"
    )


def _text(position: tuple[float, float], text: str, color: Optional[str] = None) -> str:
    size: Final = measure_latex_dimensions(text)
    x: Final = position[0] + size.width / 2.0
    y: Final = position[1] - size.height / 2.0 - size.depth / 2.0
    return rf"    \node{_options(color)} at ({x}pt, {y}pt) {{{text}}};" + "\n"


def _margin_from_text(text: str) -> float:
    text_size: Final = measure_latex_dimensions(text if text else "Placeholder")
    return text_size.total_height + _TEXT_MARGIN * 2.0


class Symbol(ABC):
    @abstractmethod
    def emit(self, position: tuple[float, float], size: tuple[float, float]) -> str:
        pass

    @property
    @abstractmethod
    def required_size(self) -> tuple[float, float]:
        pass


@final
class Imperative(Symbol):
    def __init__(self, text: str, color: Optional[str] = None) -> None:
        self._text = text
        self._color = color

    @property
    def text(self) -> str:
        return self._text

    @override
    def emit(self, position: tuple[float, float], size: tuple[float, float]) -> str:
        x1: Final = position[0]
        y1: Final = position[1]
        x2: Final = position[0] + size[0]
        y2: Final = position[1] - size[1]
        text_size: Final = measure_latex_dimensions(self._text)
        text_x: Final = position[0] + _TEXT_MARGIN
        text_y: Final = position[1] - size[1] / 2.0 + text_size.height / 2.0
        return _rectangle(x1, y1, x2, y2, self._color) + _text(
            (text_x, text_y), self._text, self._color
        )

    @override
    @property
    def required_size(self) -> tuple[float, float]:
        return (
            self._dimensions.width + 2.0 * _TEXT_MARGIN,
            self._dimensions.height + 2.0 * _TEXT_MARGIN,
        )

    @property
    def _dimensions(self) -> Dimensions:
        return measure_latex_dimensions(self._text)


@final
class Block(Symbol):
    def __init__(self, text: str, inner: Symbol) -> None:
        self._text = text
        self._inner = inner

    @property
    def text(self) -> str:
        return self._text

    @property
    def inner(self) -> Symbol:
        return self._inner

    @override
    def emit(self, position: tuple[float, float], size: tuple[float, float]) -> str:
        margin: Final = _margin_from_text(self._text)
        return (
            Imperative("").emit(position, size)
            + _text(
                (position[0] + _TEXT_MARGIN, position[1] - _TEXT_MARGIN), self._text
            )
            + self._inner.emit(
                (position[0] + margin, position[1] - margin),
                (size[0] - 2.0 * margin, size[1] - 2.0 * margin),
            )
        )

    @override
    @property
    def required_size(self) -> tuple[float, float]:
        margin: Final = _margin_from_text(self._text)
        inner_size: Final = self._inner.required_size
        return (
            inner_size[0] + 2.0 * margin,
            inner_size[1] + 2.0 * margin,
        )


@final
class Serial(Symbol):
    def __init__(self, elements: list[Symbol]) -> None:
        self._elements = elements

    @property
    def elements(self) -> list[Symbol]:
        return self._elements

    @override
    def emit(self, position: tuple[float, float], size: tuple[float, float]) -> str:
        output = ""
        element_size: Final = size[0], size[1] / len(self._elements)
        current_position = position
        for element in self._elements:
            output += element.emit(current_position, element_size)
            current_position = (
                current_position[0],
                current_position[1] - element_size[1],
            )
        return output

    @override
    @property
    def required_size(self) -> tuple[float, float]:
        max_width = 0.0
        max_height = 0.0
        for element in self._elements:
            element_size = element.required_size
            max_width = max(max_width, element_size[0])
            max_height = max(max_height, element_size[1])
        return max_width, max_height * len(self._elements)


@final
class ConstrainedSymbol(NamedTuple):
    symbol: Symbol
    min_width: float


@final
class _HorizontalSymbolBlock(Symbol):
    def __init__(
        self,
        symbol_constraints: list[ConstrainedSymbol],
        min_symbol_width: float,
    ) -> None:
        self._symbol_constraints = symbol_constraints
        self._min_symbol_width = min_symbol_width

    def get_x_of_symbol(
        self,
        base_position: tuple[float, float],
        size: tuple[float, float],
        index: int,
    ) -> float:
        additional_required_width_per_symbol: Final = max(
            0.0,
            (size[0] - self.required_size[0]) / len(self._symbol_constraints),
        )
        x = base_position[0]
        for i, symbol_constraint in enumerate(self._symbol_constraints):
            if i == index:
                return x
            x += max(
                symbol_constraint.symbol.required_size[0]
                + additional_required_width_per_symbol,
                symbol_constraint.min_width,
                self._min_symbol_width,
            )
        return x

    @override
    def emit(self, position: tuple[float, float], size: tuple[float, float]) -> str:
        result = ""
        for i, symbol_constraint in enumerate(self._symbol_constraints):
            current_position = (
                self.get_x_of_symbol(position, size, i),
                position[1],
            )
            branch_width = (
                self.get_x_of_symbol(position, size, i + 1) - current_position[0]
            )
            result += symbol_constraint.symbol.emit(
                current_position,
                (branch_width, size[1]),
            )
        return result

    @override
    @property
    def required_size(self) -> tuple[float, float]:
        required_width: Final = sum(
            max(
                constraint.symbol.required_size[0],
                constraint.min_width,
                self._min_symbol_width,
            )
            for constraint in self._symbol_constraints
        )
        return required_width, self._max_symbol_height

    @property
    def _max_symbol_height(self) -> float:
        return max(
            symbol_constraint.symbol.required_size[1]
            for symbol_constraint in self._symbol_constraints
        )


@final
class Branch(NamedTuple):
    condition: str
    inner: Symbol


class MultipleExclusiveSelective(Symbol):
    _MIN_BRANCH_WIDTH = 28.346  # 1 cm in points

    def __init__(self, common_condition_part: str, branches: list[Branch]) -> None:
        self._common_condition_part = common_condition_part
        self._branches = branches

    @property
    def common_condition_part(self) -> str:
        return self._common_condition_part

    @property
    def branches(self) -> list[Branch]:
        return self._branches

    @override
    def emit(self, position: tuple[float, float], size: tuple[float, float]) -> str:
        # Rectangle around the header.
        result = Imperative("").emit(position, (size[0], self._header_height))

        body: Final = self._body

        # Diagonal line from top left to bottom "almost" right.
        x1: Final = position[0]
        y1: Final = position[1]
        x2: Final = body.get_x_of_symbol(position, size, len(self.branches) - 1)
        y2: Final = position[1] - self._header_height

        def linear_function(x: float) -> float:
            m: Final = (y2 - y1) / (x2 - x1)
            b: Final = y1 - m * x1
            return m * x + b

        x3: Final = position[0] + size[0]
        y3: Final = position[1]
        result += _line(x1, y1, x2, y2)
        result += _line(x2, y2, x3, y3)

        # Vertical lines in the header.
        for i in range(len(self.branches) - 2):
            x = position[0] + body.get_x_of_symbol(position, size, i + 1)
            result += _line(x, linear_function(x), x, position[1] - self._header_height)

        # Condition texts in the header.
        for i in range(len(self.branches) - 1):
            text = self.branches[i].condition
            text_dimensions = measure_latex_dimensions(text)
            x_min = body.get_x_of_symbol(position, size, i) + _TEXT_MARGIN
            if len(self.branches) == 2:
                x = x_min
            else:
                x_center_of_branch = (
                    body.get_x_of_symbol(position, size, i)
                    + body.get_x_of_symbol(position, size, i + 1)
                ) / 2.0
                x_center_of_text = x_center_of_branch - text_dimensions.width / 2.0
                t = i / max(len(self.branches) - 2, 1)
                x = _lerp(x_center_of_text, x_min, t)
            result += _text(
                (
                    x,
                    position[1]
                    - self._header_height
                    + _TEXT_MARGIN
                    + text_dimensions.height,
                ),
                text,
            )

        # Rightmost condition text.
        text = self.branches[-1].condition
        text_dimensions = measure_latex_dimensions(text)
        result += _text(
            (
                position[0] + size[0] - text_dimensions.width - _TEXT_MARGIN,
                position[1]
                - self._header_height
                + _TEXT_MARGIN
                + text_dimensions.height,
            ),
            text,
        )

        # Common condition part text.
        m_x: Final = (
            position[0] + size[0] / 2.0 + position[0] + size[0] / 2.0 + x2
        ) / 3.0
        m_y: Final = (position[1] + position[1] + y2) / 3.0
        text = self.common_condition_part
        text_dimensions = measure_latex_dimensions(text)
        result += _text(
            (
                m_x - text_dimensions.width / 2.0,
                m_y + text_dimensions.height / 2.0 + text_dimensions.depth / 2.0,
            ),
            text,
        )

        # Emit branches.
        result += body.emit(
            (position[0], position[1] - self._header_height),
            (size[0], size[1] - self._header_height),
        )
        return result

    @override
    @property
    def required_size(self) -> tuple[float, float]:
        return self._total_width, self._total_height

    @property
    def _body(self) -> _HorizontalSymbolBlock:
        common_condition_width: Final = measure_latex_dimensions(
            self._common_condition_part
        ).width
        min_branch_width: Final = common_condition_width / len(self.branches) * 2.0
        return _HorizontalSymbolBlock(
            [
                ConstrainedSymbol(
                    branch.inner,
                    measure_latex_dimensions(branch.condition).width * 2.0,
                )
                for branch in self.branches
            ],
            max(
                MultipleExclusiveSelective._MIN_BRANCH_WIDTH,
                min_branch_width,
            ),
        )

    @property
    def _header_height(self) -> float:
        max_line_height: Final = max(
            measure_latex_dimensions(self.common_condition_part).total_height,
            max(
                measure_latex_dimensions(condition.condition).total_height
                for condition in self.branches
            ),
        )
        return len(self.branches) * (max_line_height + 2.0 * _TEXT_MARGIN)

    @property
    def _total_height(self) -> float:
        return self._header_height + self._body.required_size[1]

    @property
    def _total_width(self) -> float:
        return self._body.get_x_of_symbol((0.0, 0.0), (0.0, 0.0), len(self.branches))


class DyadicSelective(MultipleExclusiveSelective):
    @override
    def __init__(self, common_condition_part: str, then: Branch, else_: Branch) -> None:
        super().__init__(common_condition_part, [then, else_])


@final
class MonadicSelective(DyadicSelective):
    @override
    def __init__(self, common_condition_part: str, then: Branch) -> None:
        super().__init__(common_condition_part, then, Branch("", Imperative("")))


@final
class BlockAlignment(Enum):
    TOP = auto()
    CENTER = auto()
    BOTTOM = auto()


class Iteration(Symbol):
    def __init__(self, condition: str, body: Symbol) -> None:
        self._condition = condition
        self._body = body

    @abstractmethod
    def _block_alignment(self) -> BlockAlignment:
        pass

    @property
    def condition(self) -> str:
        return self._condition

    @property
    def body(self) -> Symbol:
        return self._body

    @override
    def emit(self, position: tuple[float, float], size: tuple[float, float]) -> str:
        margin: Final = _margin_from_text(self._condition)
        inner_height: Final = (
            size[1] - 2.0 * margin
            if self._block_alignment() == BlockAlignment.CENTER
            else size[1] - margin
        )
        inner_position_y: Final = (
            position[1]
            if self._block_alignment() == BlockAlignment.TOP
            else position[1] - margin
        )

        if self._block_alignment() == BlockAlignment.TOP:
            text_dimensions: Final = measure_latex_dimensions(self._condition)
            text_position_y = (
                position[1] - size[1] + text_dimensions.height + _TEXT_MARGIN
            )
        else:
            text_position_y = position[1] - _TEXT_MARGIN

        return (
            Imperative("").emit(position, size)
            + _text(
                (position[0] + _TEXT_MARGIN, text_position_y),
                self._condition,
            )
            + self._body.emit(
                (position[0] + margin, inner_position_y),
                (size[0] - margin, inner_height),
            )
        )

    @override
    @property
    def required_size(self) -> tuple[float, float]:
        margin: Final = _margin_from_text(self._condition)
        inner_size: Final = self.body.required_size
        return (
            inner_size[0] + margin,
            inner_size[1] + margin * 2.0
            if self._block_alignment() == BlockAlignment.CENTER
            else inner_size[1] + margin,
        )


@final
class PreTestedIteration(Iteration):
    @override
    def _block_alignment(self) -> BlockAlignment:
        return BlockAlignment.BOTTOM


@final
class PostTestedIteration(Iteration):
    @override
    def _block_alignment(self) -> BlockAlignment:
        return BlockAlignment.TOP


@final
class ContinuousIteration(Iteration):
    @override
    def __init__(self, body: Symbol) -> None:
        super().__init__("", body)

    @override
    def _block_alignment(self) -> BlockAlignment:
        return BlockAlignment.CENTER


@final
class Termination(Symbol):
    def __init__(self, text: str) -> None:
        self._text = text

    @property
    def text(self) -> str:
        return self._text

    @override
    def emit(self, position: tuple[float, float], size: tuple[float, float]) -> str:
        text_dimensions: Final = measure_latex_dimensions(self._text)
        x: Final = position[0] + self._margin
        y: Final = position[1] - size[1] / 2.0 + text_dimensions.height / 2.0
        return (
            Imperative("").emit(position, size)
            + _text((x, y), self._text)
            + _line(
                position[0] + self._margin,
                position[1],
                position[0],
                position[1] - size[1] / 2.0,
            )
            + _line(
                position[0],
                position[1] - size[1] / 2.0,
                position[0] + self._margin,
                position[1] - size[1],
            )
        )

    @override
    @property
    def required_size(self) -> tuple[float, float]:
        size: Final = Imperative(self._text).required_size
        return (
            size[0] + self._margin,
            size[1],
        )

    @property
    def _margin(self) -> float:
        return _margin_from_text("")


@final
class Parallel(Symbol):
    def __init__(self, elements: list[Symbol]) -> None:
        self._elements = elements

    @property
    def elements(self) -> list[Symbol]:
        return self._elements

    @override
    def emit(self, position: tuple[float, float], size: tuple[float, float]) -> str:
        # Upper and lower "banners".
        output = Imperative("").emit(position, (size[0], self._margin))
        output += Imperative("").emit(
            (position[0], position[1] - size[1] + self._margin),
            (size[0], self._margin),
        )

        # Diagonal lines in all four corners.
        output += _line(
            position[0],
            position[1] - self._margin,
            position[0] + self._margin,
            position[1],
        )
        output += _line(
            position[0],
            position[1] - size[1] + self._margin,
            position[0] + self._margin,
            position[1] - size[1],
        )
        output += _line(
            position[0] + size[0] - self._margin,
            position[1],
            position[0] + size[0],
            position[1] - self._margin,
        )
        output += _line(
            position[0] + size[0] - self._margin,
            position[1] - size[1],
            position[0] + size[0],
            position[1] - size[1] + self._margin,
        )

        output += self._body.emit(
            (position[0], position[1] - self._margin),
            (size[0], size[1] - 2.0 * self._margin),
        )
        return output

    @override
    @property
    def required_size(self) -> tuple[float, float]:
        body_size: Final = self._body.required_size
        return body_size[0], body_size[1] + 2.0 * self._margin

    @property
    def _margin(self) -> float:
        return _margin_from_text("")

    @property
    def _body(self) -> _HorizontalSymbolBlock:
        return _HorizontalSymbolBlock(
            [ConstrainedSymbol(element, 0.0) for element in self._elements],
            0.0,
        )
