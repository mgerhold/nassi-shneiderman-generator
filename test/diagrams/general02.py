from typing import Final

from nassi_shneiderman_generator.diagram import Diagram
from nassi_shneiderman_generator.symbols import (
    Branch,
    Block,
    Serial,
    Imperative,
    PostTestedIteration,
    ContinuousIteration,
    Termination,
    Parallel,
    MultipleExclusiveSelective,
    MonadicSelective,
    DyadicSelective,
    PreTestedIteration,
)

diagram: Final = Diagram(
    MultipleExclusiveSelective(
        "Wei√ü der Geier?",
        [
            Branch(
                "Ja",
                MonadicSelective(
                    "War geil?",
                    Branch(
                        "Ja",
                        Serial(
                            [
                                Imperative("Dann!"),
                                Imperative("Auch."),
                                Imperative("Und!"),
                            ]
                        ),
                    ),
                ),
            ),
            Branch(
                "Vielleicht",
                DyadicSelective(
                    "Wie geht's?",
                    Branch(
                        "Gut",
                        Imperative("Okay."),
                    ),
                    Branch(
                        "Schlecht",
                        Imperative("Ciao."),
                    ),
                ),
            ),
            Branch(
                "Nein",
                PreTestedIteration(
                    r"$\texttt{n} < 10$",
                    Serial(
                        [
                            Imperative(r"Ausgabe: \texttt{n}"),
                            Imperative(r"$\texttt{n} := \texttt{n} + 1$"),
                        ]
                    ),
                ),
            ),
            Branch(
                "Blubb",
                PostTestedIteration(
                    r"$\texttt{n} < 10$",
                    Serial(
                        [
                            Imperative(r"Ausgabe: \texttt{n}"),
                            Imperative(r"$\texttt{n} := \texttt{n} + 1$"),
                        ]
                    ),
                ),
            ),
            Branch(
                "Huh?",
                ContinuousIteration(
                    Serial(
                        [
                            Imperative(r"Ausgabe: \texttt{n}"),
                            Imperative(r"$\texttt{n} := \texttt{n} + 1$"),
                            Termination("Raus hier!"),
                            Parallel(
                                [
                                    Imperative("a"),
                                    Imperative("b"),
                                    Imperative("c"),
                                ]
                            ),
                        ]
                    ),
                ),
            ),
            Branch(
                "Nein",
                Block(
                    "Mein Block",
                    Serial(
                        [
                            Imperative(r"Eingabe: \texttt{n}"),
                            Imperative(r"$\texttt{n} := \texttt{n} + 1$"),
                            Imperative(r"Was geht ab? $\frac{1}{2}$"),
                            Imperative(r"a"),
                            Imperative(r"-"),
                        ]
                    ),
                ),
            ),
        ],
    ),
)
