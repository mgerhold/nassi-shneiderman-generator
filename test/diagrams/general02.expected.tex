\begin{tikzpicture}
    \draw (0pt,0pt) rectangle (540.77969pt, -89.66664pt);
    \node at (4.0pt, -44.83332pt) {};
    \draw (0pt, 0pt) -- (433.11571pt, -89.66664pt);
    \draw (433.11571pt, -89.66664pt) -- (540.77969pt, 0pt);
    \draw (82.83346pt, -17.148761557908855pt) -- (82.83346pt, -89.66664pt);
    \draw (205.38916pt, -42.52109873738452pt) -- (205.38916pt, -89.66664pt);
    \draw (277.97463pt, -57.54824981375808pt) -- (277.97463pt, -89.66664pt);
    \draw (350.5601pt, -72.57540089013165pt) -- (350.5601pt, -89.66664pt);
    \node at (41.41673pt, -82.24998500000001pt) {Ja};
    \node at (134.89602125pt, -82.19442000000001pt) {Vielleicht};
    \node at (230.6049825pt, -82.24998500000001pt) {Nein};
    \node at (299.99576125pt, -82.19442000000001pt) {Blubb};
    \node at (366.226795pt, -82.19442000000001pt) {Huh?};
    \node at (526.64078pt, -82.24998500000001pt) {Nein};
    \node at (324.6317999999999pt, -29.88888pt) {Wei√ü der Geier?};
    \draw (0pt,-89.66664pt) rectangle (82.83346pt, -123.4444pt);
    \node at (4.0pt, -106.55552pt) {};
    \draw (0pt, -89.66664pt) -- (41.41673pt, -123.4444pt);
    \draw (41.41673pt, -123.4444pt) -- (82.83346pt, -89.66664pt);
    \node at (9.06946pt, -116.02774500000001pt) {Ja};
    \node at (78.83346pt, -119.4444pt) {};
    \node at (41.41673pt, -100.92589333333333pt) {War geil?};
    \draw (0pt,-123.4444pt) rectangle (41.41673pt, -181.92586666666668pt);
    \node at (17.263925pt, -152.68513333333334pt) {Dann!};
    \draw (0pt,-181.92586666666668pt) rectangle (41.41673pt, -240.40733333333336pt);
    \node at (16.7778pt, -211.16660000000002pt) {Auch.};
    \draw (0pt,-240.40733333333336pt) rectangle (41.41673pt, -298.8888pt);
    \node at (14.694475pt, -269.6480666666667pt) {Und!};
    \draw (41.41673pt,-123.4444pt) rectangle (82.83346pt, -298.8888pt);
    \node at (45.41673pt, -211.16660000000002pt) {};
    \draw (82.83346pt,-89.66664pt) rectangle (205.38916pt, -123.4444pt);
    \node at (86.83346pt, -106.55552pt) {};
    \draw (82.83346pt, -89.66664pt) -- (133.72242pt, -123.4444pt);
    \draw (133.72242pt, -123.4444pt) -- (205.38916pt, -89.66664pt);
    \node at (95.47931pt, -116.02774500000001pt) {Gut};
    \node at (183.472475pt, -115.97218000000001pt) {Schlecht};
    \node at (140.64834666666667pt, -100.92589333333333pt) {Wie geht's?};
    \draw (82.83346pt,-123.4444pt) rectangle (133.72242pt, -298.8888pt);
    \node at (99.05572000000001pt, -212.13882pt) {Okay.};
    \draw (133.72242pt,-123.4444pt) rectangle (205.38916pt, -298.8888pt);
    \node at (149.111335pt, -211.16660000000002pt) {Ciao.};
    \draw (205.38916pt,-89.66664pt) rectangle (277.97463pt, -298.8888pt);
    \node at (209.38916pt, -194.27772pt) {};
    \node at (223.68076pt, -97.08435pt) {$\texttt{n} < 10$};
    \draw (220.22458pt,-104.50206pt) rectangle (277.97463pt, -201.69543pt);
    \node at (249.099605pt, -154.070965pt) {Ausgabe: \texttt{n}};
    \draw (220.22458pt,-201.69543pt) rectangle (277.97463pt, -298.8888pt);
    \node at (246.141115pt, -250.70878pt) {$\texttt{n} := \texttt{n} + 1$};
    \draw (277.97463pt,-89.66664pt) rectangle (350.5601pt, -298.8888pt);
    \node at (281.97463pt, -194.27772pt) {};
    \node at (296.26623pt, -291.86207pt) {$\texttt{n} < 10$};
    \draw (292.81005pt,-89.66664pt) rectangle (350.5601pt, -186.86001pt);
    \node at (321.685075pt, -139.235545pt) {Ausgabe: \texttt{n}};
    \draw (292.81005pt,-186.86001pt) rectangle (350.5601pt, -284.05338pt);
    \node at (318.726585pt, -235.87336pt) {$\texttt{n} := \texttt{n} + 1$};
    \draw (350.5601pt,-89.66664pt) rectangle (433.11571pt, -298.8888pt);
    \node at (354.5601pt, -194.27772pt) {};
    \node at (354.5601pt, -93.66664pt) {};
    \draw (365.50453999999996pt,-104.61108pt) rectangle (433.11571pt, -149.4444pt);
    \node at (394.37956499999996pt, -127.99995999999999pt) {Ausgabe: \texttt{n}};
    \draw (365.50453999999996pt,-149.4444pt) rectangle (433.11571pt, -194.27772pt);
    \node at (391.421075pt, -172.277725pt) {$\texttt{n} := \texttt{n} + 1$};
    \draw (365.50453999999996pt,-194.27772pt) rectangle (433.11571pt, -239.11104pt);
    \node at (369.50453999999996pt, -216.69438pt) {};
    \node at (402.78234499999996pt, -216.69438pt) {Raus hier!};
    \draw (380.44897999999995pt, -194.27772pt) -- (365.50453999999996pt, -216.69438pt);
    \draw (365.50453999999996pt, -216.69438pt) -- (380.44897999999995pt, -239.11104pt);
    \draw (365.50453999999996pt,-239.11104pt) rectangle (433.11571pt, -254.05548pt);
    \node at (369.50453999999996pt, -246.58326pt) {};
    \draw (365.50453999999996pt,-268.99992000000003pt) rectangle (433.11571pt, -283.94436pt);
    \node at (369.50453999999996pt, -276.47214pt) {};
    \draw (365.50453999999996pt, -254.05548pt) -- (380.44897999999995pt, -239.11104pt);
    \draw (365.50453999999996pt, -268.99992000000003pt) -- (380.44897999999995pt, -283.94436pt);
    \draw (418.17127pt, -239.11104pt) -- (433.11571pt, -254.05548pt);
    \draw (418.17127pt, -283.94436pt) -- (433.11571pt, -268.99992000000003pt);
    \draw (365.50453999999996pt,-254.05548pt) rectangle (388.0416066666666pt, -268.99992pt);
    \node at (372.00454999999994pt, -261.5277pt) {a};
    \draw (388.0416066666666pt,-254.05548pt) rectangle (411.1342233333333pt, -268.99992pt);
    \node at (394.8193916666666pt, -261.5277pt) {b};
    \draw (411.1342233333333pt,-254.05548pt) rectangle (433.11571pt, -268.99992pt);
    \node at (417.3564433333333pt, -261.5277pt) {c};
    \draw (433.11571pt,-89.66664pt) rectangle (540.77969pt, -298.8888pt);
    \node at (437.11571pt, -194.27772pt) {};
    \node at (462.04632pt, -97.13886pt) {Mein Block};
    \draw (448.06014999999996pt,-104.61108pt) rectangle (525.83525pt, -140.477736pt);
    \node at (476.004625pt, -123.516628pt) {Eingabe: \texttt{n}};
    \draw (448.06014999999996pt,-140.477736pt) rectangle (525.83525pt, -176.344392pt);
    \node at (473.976685pt, -158.82772899999998pt) {$\texttt{n} := \texttt{n} + 1$};
    \draw (448.06014999999996pt,-176.344392pt) rectangle (525.83525pt, -212.211048pt);
    \node at (486.94769999999994pt, -196.001925pt) {Was geht ab? $\frac{1}{2}$};
    \draw (448.06014999999996pt,-212.211048pt) rectangle (525.83525pt, -248.077704pt);
    \node at (454.56015999999994pt, -230.144376pt) {a};
    \draw (448.06014999999996pt,-248.077704pt) rectangle (525.83525pt, -283.94436pt);
    \node at (453.726815pt, -266.011032pt) {-};
\end{tikzpicture}
